### PR TITLE
feat: pinning services stage 1

### PR DESCRIPTION
**Pull request by rafaelramalho19**
This is a pull request that marks the stage 1 of pinning services.

This includes:
 - Local pinning now works with future-proof modal for remote pins
 - Settings page now reflects how much space local pinning occupies.

Closes https://github.com/ipfs-shipyard/ipfs-webui/issues/1500
Closes https://github.com/ipfs-shipyard/ipfs-webui/issues/1501

### CODE REVIEW

File: src/components/pinning-manager/PinningManager.js
Comment (jessicaschilling) [inline src/components/pinning-manager/PinningManager.js:53]:
```suggestion
        <div className='ph3 flex items-center bg-white lh-copy charcoal f6 fw5'>
```
Looks a little more consistent, and gives us a tiny bit more space at narrower widths.

File: src/components/pinning-manager/PinningManager.js
Comment (jessicaschilling) [inline src/components/pinning-manager/PinningManager.js:96]:
```suggestion
  const iconClass = 'mr2 pr1 fill-aqua'
```
Color consistency was my fault, sorry :( 

Also, can the icon please not scale as the page width becomes narrower?

File: src/components/modal/Modal.js
Comment (jessicaschilling) [inline src/components/modal/Modal.js:22]:
```suggestion
  <div className={`ph4 pv3 tc ${className}`} {...props}>
```
To avoid ugly margins.

File: src/files/modals/pinning-modal/PinningModal.js
Comment (jessicaschilling) [inline src/files/modals/pinning-modal/PinningModal.js:65]: Can we please add an `app:actions.apply` key (if there isn't one already) per @lidel's suggestion for language for this button?

File: public/locales/en/files.json
Comment (jessicaschilling) [inline public/locales/en/files.json:41]: Seeking thoughts from @rafaelramalho19 and @lidel here ... It's a little weird to not offer any explanatory context for "why do I only have one choice and what's a service??" in this modal:
image
That said, I think we're probably OK to leave as-is if this is only (a) a short-term situation and (b) only visible to early adopters loading master from webui.ipfs.io. Your thoughts?

**Review by jessicaschilling — CHANGES_REQUESTED** [review]
Thank you! 😊  Functionally this seems perfect on my end. Left some comments inline on visual/UX stuff, but a few other thoughts:

1. On the Settings screen, the left-most table items (both headers and row cells) have both a 10px `margin-left` and a `pl2` on them. This has the effect of the leftmost column of the table looking too indented as a whole. Not sure which styling should be removed to mitigate this, though?

2. On the Settings screen table, can the rows' cells truncate with ellipses at narrow widths (like the headers do), and can their full values display in hover text (also like the headers)?

3. We'll need to add explanatory content to the CLI tutor modal to this effect: `This command will only pin this item's CID to your local node. If you want to persist it with a remote pinning service, consult that service's CLI documentation to learn how.` Right now we have i18n key `cliModal.extraNotes` for that purpose, but it's effectively hard-coded for one use (the config in Settings). The best way to remedy might be to move `cliModal.extraNotes` out of `app.json` and into `settings.json` and `files.json`?

File: public/locales/en/settings.json
Comment (rafaelramalho19) [reply public/locales/en/settings.json:8]: This is not present in the current release tho, that can be changed in the `epic/pinning-services` branch

File: src/components/modal/Modal.js
Comment (rafaelramalho19) [reply src/components/modal/Modal.js:22]: Warning: This will change all modals

File: public/locales/en/files.json
Comment (rafaelramalho19) [reply public/locales/en/files.json:41]: We can even have a release with this. It just transitions the users to the new flow, and in the patch notes they can keep track of what's being changed too.

Don't think it's necessary to specify why we only have a service for now, I even think it smoothens the transition between local pinning only into remote pinning 😄

File: src/components/pinning-manager/PinningManager.js
Comment (rafaelramalho19) [reply src/components/pinning-manager/PinningManager.js:96]: Done!

File: public/locales/en/settings.json
Comment (rafaelramalho19) [reply public/locales/en/settings.json:8]: Oh I missread what you said, sorry. I will change it for now then :)

**Comment by rafaelramalho19** [comment]
> 3. We'll need to add explanatory content to the CLI tutor modal to this effect: `This command will only pin this item's CID to your local node. If you want to persist it with a remote pinning service, consult that service's CLI documentation to learn how.` Right now we have i18n key `cliModal.extraNotes` for that purpose, but it's effectively hard-coded for one use (the config in Settings). The best way to remedy might be to move `cliModal.extraNotes` out of `app.json` and into `settings.json` and `files.json`?

I think that's a good point, but it's something that doesn't need to be addressed in this stage :)

File: src/components/modal/Modal.js
Comment (jessicaschilling) [reply src/components/modal/Modal.js]: Yep, I'll check them all.

**Comment by jessicaschilling** [comment]
@rafaelramalho19 Added https://github.com/ipfs-shipyard/ipfs-webui/issues/1686 for the CLI tutor mode change - I'll do that real quick once this issue gets merged.

File: src/files/header/Header.js
Comment (jessicaschilling) [inline src/files/header/Header.js:16]:
```suggestion
    <span className={`nowrap db ttu f6 montserrat fw4 ${isLink ? 'link' : 'charcoal-muted'}`}>{title}</span>
```

This should take care of the visual wierdness mentioned in https://github.com/ipfs-shipyard/ipfs-webui/pull/1685#discussion_r517561452 ?

**Review by jessicaschilling — CHANGES_REQUESTED** [review]
LGTM, just one thing that got left out from the original review 😊 

On the Settings screen table, can the rows' cells truncate with ellipses at narrow widths (like the headers do), and can their full values display in hover text (also like the headers)?

File: src/files/header/Header.js
Comment (rafaelramalho19) [reply src/files/header/Header.js:16]: Oh sorry, missed that one! Good catch 👍

**Review by jessicaschilling — CHANGES_REQUESTED** [review]
On the Settings screen table, can the rows' cells truncate with ellipses at narrow widths (like the headers do), and can their full values display in hover text (also like the headers)?

Otherwise this happens:

image

**Review by lidel — CHANGES_REQUESTED** [review]
Wrote down some suggestions and one minor blocker inline, but hopefully all those are easy to fix.
Apart from that, looks wonderful! :+1: 

ps. Some users may be impacted by the fact this PR removes old `#/pins` screen, and with that, the ability to remove a pin after an item is removed from Files. I don't think this is a problem as long we land ability to `Delete and Unpin` (https://github.com/ipfs-shipyard/ipfs-webui/pull/1644). Just mentioning here, so we don't forget.

ps2. Does this supersede  #1615? Should I close that PR?

File: public/locales/en/files.json
Comment (jessicaschilling) [reply public/locales/en/files.json:41]: Let's go with the following (note, replaced "wish" with "would like" so it doesn't sound like wishing on magic beans, and risk weird translations):

```suggestion
    "title": "Select where you would like to pin these items.",
```

File: src/components/pinning-manager/PinningManager.js
Comment (jessicaschilling) [reply src/components/pinning-manager/PinningManager.js:71]: @lidel Good memory - thank you for calling that out!
Agree that indicating the number of items pinned is the best choice, particularly since we're removing that from the info bar in the Files screen. However, I'd suggest just `Pins` for the column title so it doesn't truncate super-easily (IMO the fact that there is no unit in the cells for the Pins column, as opposed to "MB" units for the Files column, implies quantity without any need for further explanation).

File: src/components/pinning-manager/PinningManager.js
Comment (jessicaschilling) [reply src/components/pinning-manager/PinningManager.js:70]: Agree that column header should be `Size` - plus that's consistent with how things are labeled in the Files screen.

File: public/locales/en/settings.json
Comment (jessicaschilling) [reply public/locales/en/settings.json:8]: How about this? Makes it feel a little more like insurance, i.e. "pin stuff you really want to keep just in case you accidentally delete it from MFS."

```suggestion
    "description": "Support for third-party remote pinning services will be implemented in a future release, and you'll be able to add and configure them here. In the meantime, use local pinning when you want to ensure an item on your node is never garbage-collected, even if you remove it from Files."
```

File: src/files/header/Header.js
Comment (jessicaschilling) [reply src/files/header/Header.js:64]: How about the following, so it makes sense both in your top-level directory and lower-level ones?
`Total size of data in the current directory (if a subdirectory, the size of all data in Files is also displayed)`

File: src/files/header/Header.js
Comment (jessicaschilling) [reply src/files/header/Header.js:74]: How about this?
`Total size of all blocks in your IPFS repo; this includes all contents of Files, plus all locally pinned items and any temporary cached data`

**Comment by rafaelramalho19** [comment]
> Wrote down some suggestions and one minor blocker inline, but hopefully all those are easy to fix.
> Apart from that, looks wonderful! 👍
> 
> ps. Some users may be impacted by the fact this PR removes old `#/pins` screen, and with that, the ability to remove a pin after an item is removed from Files. I don't think this is a problem as long we land ability to `Delete and Unpin` (#1644). Just mentioning here, so we don't forget.
> 
> ps2. Does this supersede #1615? Should I close that PR?

Good call, closing it 👍

File: src/components/pinning-manager/PinningManager.js
Comment (rafaelramalho19) [reply src/components/pinning-manager/PinningManager.js:70]: ✅

File: src/files/header/Header.js
Comment (lidel) [reply src/files/header/Header.js:64]: Hm.. I think you have a point, but i dont think clarification in title is enough to remove confusion in subdirectories ("why FILES got smaller, but ALL BLOCKS remained the same??")

Maybe we should just hide "All BLOCKS" when in subdirectory for now?

Comment (jessicaschilling) [reply src/files/header/Header.js]: @lidel - The "also displayed" refers to the gray smaller text, not the quantity in "All Blocks" -- so I'd say this is OK to leave as suggested?
image

File: public/locales/en/files.json
Comment (jessicaschilling) [inline public/locales/en/files.json:102]:
```suggestion
  "allBlocksDescription": "Total size of blocks on your entire IPFS node; this includes everything in Files, plus all locally pinned items and any temporary cached data",
```

**Review by jessicaschilling — CHANGES_REQUESTED** [review]

**Review by jessicaschilling — APPROVED** [review]

File: src/files/header/Header.js
Comment (jessicaschilling) [reply src/files/header/Header.js]: Resolved by improving helptext for "All Blocks" in https://github.com/ipfs-shipyard/ipfs-webui/pull/1685/commits/992d693da69a05473066125f6a15af9603cbbea3

File: src/components/pinning-manager/PinningManager.js
Comment (rafaelramalho19) [reply src/components/pinning-manager/PinningManager.js]: Forgot about this, thank you 🙏

File: src/components/pinning-manager/PinningManager.js
Comment (rafaelramalho19) [reply src/components/pinning-manager/PinningManager.js:70]: Done ✅

**Review by lidel — CHANGES_REQUESTED** [review]
@rafaelramalho19 seems e2e tests started failing after you merged updates from `master` – let me know if you are looking into it or got blocked and need a second pair of eyes.

**Comment by lidel** [comment]
@rafaelramalho19 I was able to reproduce the CI error locally on Linux with: 

```
npm ci && npm run build && npm run test:e2e
```

Note that E2E require prebuilt static version. Maybe you were running against older build? Mind checking?

Unfortunately I've run out of time to look deeper into this. Will pick up on thursday if still needed.

**Comment by rafaelramalho19** [comment]
> @rafaelramalho19 I was able to reproduce the CI error locally on Linux with:
> 
> ```
> npm ci && npm run build && npm run test:e2e
> ```
> 
> Note that E2E test suite requires a prebuilt static version.
> Maybe you were running against older build? Mind checking?
> 
> What is interesting is that the tests _do not fail_ when you run debug mode (debug runs tests in slowed down mode and with browser window visible):
> 
> ```
> DEBUG=true npm run test:e2e
> ```
> 
> If you disable `sloMo` then in GUI can see a problem when JSON config is quickly typed into the field – sometimes it becomes garbled:
> 
> > 
> 
> If I set `sloMo` to `50` then headless tests are green.
> Looks like some kind of regression, not sure what caused it.
> 
> Unfortunately I've run out of time to look deeper into this and I'm OOO on Wednesday – will pick up on Thursday if still needed.

Weirdly it doesn't happen in master, so I guess it was due to changes in the settings page in this branch 🤔

**Comment by lidel** [comment]
PSA: I am helping with https://github.com/ipfs/go-ipfs/pull/7661 won't be able to look into this until next week.

**Review by lidel — APPROVED** [review]
@rafaelramalho19 I fixed E2E – details in 7089d93 (better late than never :sweat_smile:)

Let's discuss next steps during Monday's issue triage,
but if you start work before the call, my quick suggestion is to focus on steps 1 & 2:
1. Merge this PR – LGTM! :heart: 
   - nit: calculating total size of all pins is expensive: I have over 1k of pins and it takes minutes of "Loading..." so its pretty useless for users with a lot of pins. That being said, it is  not a blocker, it runs in background, and we can revisit/optimize in (2)
2. Start a new PR to wire up Pinning Services
   - while you were OoO go-ipfs 0.8.0-rc1 shipped and Irakli wired things up in https://github.com/ipfs/js-ipfs/pull/3293
     - use `ipfs-http-client` from that PR for now
     - for quick reference, API docs for `ipfs.pin.remote` commands: https://github.com/ipfs/js-ipfs/blob/feat/pin-remote/docs/core-api/PIN.md
3. (TBD) tag a release without (2) 
   - We should tag an intermediate release to be included in go-ipfs 0.8.0 – open a PR and use version from this PR, and if time allows then switch 0.8.0 to version with Pinning Services, if not, we would ship go-ipfs without it and add it in January in 0.8.1
   - As this PR removes screen with low level pins, we kinda need to rebase and finish https://github.com/ipfs-shipyard/ipfs-webui/pull/1644 to give users means of removing pins while they remove files from

**Pull request merged**
